plugins/filter_aws/aws.c

```diff
@@ -182,191 +203,250 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
+    if (ctx->tag_keys) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_keys[i]) {
+                flb_sds_destroy(ctx->tag_keys[i]);
+            }
         }
-        flb_http_client_destroy(client);
-        flb_upstream_conn_release(u_conn);
-        return -1;
+        flb_free(ctx->tag_keys);
+        ctx->tag_keys = NULL;
+    }
+    if (ctx->tag_values) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_values[i]) {
+                flb_sds_destroy(ctx->tag_values[i]);
+            }
+        }
+        flb_free(ctx->tag_values);
+        ctx->tag_values = NULL;
     }
-
-    ctx->imds_v2_token = flb_sds_create_len(client->resp.payload,
-                                            client->resp.payload_size);
-    if (!ctx->imds_v2_token) {
-        flb_errno();
-        flb_http_client_destroy(client);
-        flb_upstream_conn_release(u_conn);
-        return -1;
+    if (ctx->tag_keys_len) {
+        flb_free(ctx->tag_keys_len);
     }
-    ctx->imds_v2_token_len = client->resp.payload_size;
-
-    flb_http_client_destroy(client);
-    flb_upstream_conn_release(u_conn);
-    return 0;
+    ctx->tag_keys_len = NULL;
+    if (ctx->tag_values_len) {
+        flb_free(ctx->tag_values_len);
+    }
+    ctx->tag_values_len = NULL;
+    ctx->tags_count = 0;
 }
 
-/* get the metadata by key if the result is a json object.
- * If key is NULL, just return the value it get.
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
  */
-static int get_metadata_by_key(struct flb_filter_aws *ctx, char *metadata_path,
-                               flb_sds_t *metadata, size_t *metadata_len,
-                               char *key)
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
 {
     int ret;
-    size_t b_sent;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    flb_sds_t tag_key;
     flb_sds_t tmp;
-    struct flb_http_client *client;
-    struct flb_connection *u_conn;
+    size_t tag_key_len;
+    int i;
 
-    u_conn = flb_upstream_conn_get(ctx->ec2_upstream);
-    if (!u_conn) {
-        flb_plg_error(ctx->ins, "connection initialization error");
+    /* get a list of tag keys from the meta data server */
+    ret = flb_aws_imds_request(ctx->client_imds, FLB_AWS_IMDS_INSTANCE_TAG, &tags_list,
+							   &len);
+    if (ret < 0) {
+        ctx->tags_count = 0;
+        if (ret == -2) { /* if there are no tags, response status code is 404 */
+            flb_plg_warn(ctx->ins, "EC2 instance metadata tag request returned 404. "
+                                   "This likely indicates your instance has no tags "
+                                   "or the EC2 tagging metadata API is not enabled");
+            return 0;
+        }
+        flb_sds_destroy(tags_list);
```

Comment:
So the behavior here is that if we get a 404, we don't retry to fetch tags, and just continue on with other metadata. I think this is a good solution.

However, its implemented currently by having this function return success for this case, and set the `ctx->tags_count` to 0, which makes it so that `get_ec2_tag_values` does nothing when its called, and then `ctx->tags_fetched = FLB_TRUE;` is set in `get_ec2_tags`.

While this worked, I think the logic is not very readable and could easily be broken by a future change by an engineer who didn't understand this. I'd prefer if you explicitly handle this case.

Something like, return a failure value that's differently than -1, may be -2 again, and set `ctx->tags_fetched = FLB_TRUE;` in the calling function for this case. And add a nice comment.

Reply:
> While this worked, I think the logic is not very readable and could easily be broken by a future change by an engineer who didn't understand this. I'd prefer if you explicitly handle this case.

> Something like, return a failure value that's differently than -1, may be -2 again, and set ctx->tags_fetched = FLB_TRUE; in the calling function for this case. And add a nice comment.

I've applied your suggestion and handle the 404 / no tags explicitly now.

Reply:
Thanks, let me test this out and then I think we should be good to go.